Index: libccd/configure.ac
===================================================================
--- libccd/configure.ac	(revision 1771)
+++ libccd/configure.ac	(working copy)
@@ -5,6 +5,7 @@
 AC_INIT([libccd], [1.0], [danfis@danfis.cz])
 AC_CONFIG_SRCDIR([src/ccd.c])
 AC_CONFIG_HEADERS([src/ccd/config.h])
+AC_CONFIG_MACRO_DIR([m4])
 AM_INIT_AUTOMAKE
 
 # Checks for programs.
Index: libccd/Makefile.am
===================================================================
--- libccd/Makefile.am	(revision 1771)
+++ libccd/Makefile.am	(working copy)
@@ -1,5 +1,7 @@
 SUBDIRS = src
 
+ACLOCAL_AMFLAGS = -I m4
+
 EXTRA_DIST = \
 	BSD-LICENSE \
 	README
